File: vane-core/src/main/java/org/oddlama/vane/util/BlockUtil.java
package org.oddlama.vane.util;

import static org.oddlama.vane.util.MaterialUtil.isReplaceableGrass;
import static org.oddlama.vane.util.MaterialUtil.isTillable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import net.minecraft.core.BlockPos;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.Skull;
import org.bukkit.craftbukkit.CraftWorld;
import org.bukkit.entity.LivingEntity;
import org.bukkit.inventory.ItemStack;
import org.bukkit.util.BlockVector;
import org.bukkit.util.Vector;
import org.jetbrains.annotations.NotNull;

public class BlockUtil {

    public static final List<BlockFace> BLOCK_FACES = Arrays.asList(
        BlockFace.NORTH,
        BlockFace.EAST,
        BlockFace.SOUTH,
        BlockFace.WEST,
        BlockFace.UP,
        BlockFace.DOWN
    );

    public static final List<BlockFace> XZ_FACES = Arrays.asList(
        BlockFace.NORTH,
        BlockFace.EAST,
        BlockFace.SOUTH,
        BlockFace.WEST
    );

    public static final int NEAREST_RELATIVE_BLOCKS_FOR_RADIUS_MAX = 6;
    public static final List<List<BlockVector>> NEAREST_RELATIVE_BLOCKS_FOR_RADIUS = new ArrayList<>();

    static {
        for (int i = 0; i <= NEAREST_RELATIVE_BLOCKS_FOR_RADIUS_MAX; ++i) {
            NEAREST_RELATIVE_BLOCKS_FOR_RADIUS.add(nearestBlocksForRadius(i));
        }
    }

    public static boolean equalsPos(final Block b1, final Block b2) {
        return b1.getX() == b2.getX() && b1.getY() == b2.getY() && b1.getZ() == b2.getZ();
    }

    public static void dropNaturally(Block block, ItemStack drop) {
        dropNaturally(block.getLocation().add(0.5, 0.5, 0.5), drop);
    }

    public static void dropNaturally(Location loc, ItemStack drop) {
        loc
            .getWorld()
            .dropItem(loc.add(Vector.getRandom().subtract(new Vector(.5, .5, .5)).multiply(0.5)), drop)
            .setVelocity(Vector.getRandom().add(new Vector(-.5, +.5, -.5)).normalize().multiply(.15));
    }

    public static List<BlockVector> nearestBlocksForRadius(int radius) {
        final var ret = new ArrayList<BlockVector>();

        // Use square bounding box
        for (int x = -radius; x <= radius; x++) {
            for (int z = -radius; z <= radius; z++) {
                // Only circular area
                if (x * x + z * z > radius * radius + 0.5) {
                    continue;
                }

                ret.add(new BlockVector(x, 0, z));
            }
        }

        Collections.sort(ret, new BlockVectorRadiusComparator());
        return ret;
    }

    public static @NotNull Block relative(@NotNull final Block block, @NotNull final Vector relative) {
        return block.getRelative(relative.getBlockX(), relative.getBlockY(), relative.getBlockZ());
    }

    public static Block nextTillableBlock(final Block rootBlock, int radius, boolean careless) {
        for (final var relativePos : NEAREST_RELATIVE_BLOCKS_FOR_RADIUS.get(radius)) {
            final var block = relative(rootBlock, relativePos);

            // Check for a tillable material
            if (!isTillable(block.getType())) {
                continue;
            }

            // Get block above
            final var above = block.getRelative(0, 1, 0);

            if (above.getType() == Material.AIR) {
                // If the block above is air, we can till the block.
                return block;
            } else if (careless && isReplaceableGrass(above.getType())) {
                // If the block above is replaceable grass, delete it and return the block.
                above.setType(Material.AIR);
                return block;
            }
        }

        // We are outside the radius.
        return null;
    }

    public static Block[] adjacentBlocks3D(final Block root) {
        final var adjacent = new Block[26];

        // Direct adjacent
        adjacent[0] = root.getRelative(1, 0, 0);
        adjacent[1] = root.getRelative(-1, 0, 0);
        adjacent[2] = root.getRelative(0, 0, 1);
        adjacent[3] = root.getRelative(0, 0, -1);
        adjacent[4] = root.getRelative(0, 1, 0);
        adjacent[5] = root.getRelative(0, -1, 0);

        // Edge adjacent
        adjacent[6] = root.getRelative(0, -1, -1);
        adjacent[7] = root.getRelative(0, -1, 1);
        adjacent[8] = root.getRelative(0, 1, -1);
        adjacent[9] = root.getRelative(0, 1, 1);
        adjacent[10] = root.getRelative(-1, 0, -1);
        adjacent[11] = root.getRelative(-1, 0, 1);
        adjacent[12] = root.getRelative(1, 0, -1);
        adjacent[13] = root.getRelative(1, 0, 1);
        adjacent[14] = root.getRelative(-1, -1, 0);
        adjacent[15] = root.getRelative(-1, 1, 0);
        adjacent[16] = root.getRelative(1, -1, 0);
        adjacent[17] = root.getRelative(1, 1, 0);

        // Corner adjacent
        adjacent[18] = root.getRelative(-1, -1, -1);
        adjacent[19] = root.getRelative(-1, -1, 1);
        adjacent[20] = root.getRelative(-1, 1, -1);
        adjacent[21] = root.getRelative(-1, 1, 1);
        adjacent[22] = root.getRelative(1, -1, -1);
        adjacent[23] = root.getRelative(1, -1, 1);
        adjacent[24] = root.getRelative(1, 1, -1);
        adjacent[25] = root.getRelative(1, 1, 1);

        return adjacent;
    }

    public static Block nextSeedableBlock(final Block rootBlock, Material farmlandType, int radius) {
        for (var relativePos : NEAREST_RELATIVE_BLOCKS_FOR_RADIUS.get(radius)) {
            final var block = relative(rootBlock, relativePos);
            final var below = block.getRelative(BlockFace.DOWN);

            // The Block below must be farmland and the block itself must be air
            if (below.getType() == farmlandType && block.getType() == Material.AIR) {
                return block;
            }
        }

        // We are outside the radius.
        return null;
    }

    public static void updateLever(final Block block, final BlockFace face) {
        final var level = ((CraftWorld) block.getWorld()).getHandle();
        final var connectedBlock = block.getRelative(face.getOppositeFace());
        final var blockPos1 = new BlockPos(block.getX(), block.getY(), block.getZ());
        final var blockPos2 = new BlockPos(connectedBlock.getX(), connectedBlock.getY(), connectedBlock.getZ());
        final var initiatorBlock = level.getBlockIfLoaded(blockPos1);
        level.updateNeighborsAt(blockPos1, initiatorBlock);
        level.updateNeighborsAt(blockPos2, initiatorBlock);
    }

    public static class Corner {

        private boolean x;
        private boolean y;
        private boolean z;
        private int id;

        public Corner(final Vector hit) {
            this(hit.getX() >= 0.0, hit.getY() >= 0.0, hit.getZ() >= 0.0);
        }

        public Corner(boolean x, boolean y, boolean z) {
            this.x = x;
            this.y = y;
            this.z = z;
            final var mx = x ? 1 : 0;
            final var my = y ? 1 : 0;
            final var mz = z ? 1 : 0;
            this.id = (mx << 2) | (my << 1) | (mz << 0);
        }

        public boolean up() {
            return y;
        }

        public boolean east() {
            return x;
        }

        public boolean south() {
            return z;
        }

        public Corner up(boolean up) {
            return new Corner(x, up, z);
        }

        public Corner east(boolean east) {
            return new Corner(east, y, z);
        }

        public Corner south(boolean south) {
            return new Corner(x, y, south);
        }

        /** Rotates the corner as if it were on a north facing block given the block's rotation. */
        public Corner rotateToNorthReference(final BlockFace rotation) {
            switch (rotation) {
                default:
                    throw new IllegalArgumentException("rotation must be one of NORTH, EAST, SOUTH, WEST");
                case NORTH:
                    return new Corner(x, y, z);
                case EAST:
                    return new Corner(z, y, !x);
                case SOUTH:
                    return new Corner(!x, y, !z);
                case WEST:
                    return new Corner(!z, y, x);
            }
        }

        /** Returns {NORTH, SOUTH}_{EAST, WEST} to indicate the XZ corner. */
        public BlockFace xzFace() {
            if (x) {
                if (z) {
                    return BlockFace.SOUTH_EAST;
                } else {
                    return BlockFace.NORTH_EAST;
                }
            } else {
                if (z) {
                    return BlockFace.SOUTH_WEST;
                } else {
                    return BlockFace.NORTH_WEST;
                }
            }
        }

        @Override
        public int hashCode() {
            return id;
        }

        @Override
        public boolean equals(Object other) {
            return other instanceof Corner && ((Corner) other).id == id;
        }
    }

    public static class Oct {

        private Vector hitPos; // Relative to a block middle
        private Corner corner;
        private BlockFace face;

        public Oct(final Vector hitPos, final BlockFace face) {
            this.hitPos = hitPos;
            this.corner = new Corner(hitPos);
            this.face = face;
        }

        public Vector hitPos() {
            return hitPos;
        }

        public Corner corner() {
            return corner;
        }

        public BlockFace face() {
            return face;
        }
    }

    public static Oct raytraceOct(final LivingEntity entity, final Block block) {
        // Ray-trace position and face
        final var result = entity.rayTraceBlocks(10.0);
        if (block == null || result == null || !block.equals(result.getHitBlock())) {
            return null;
        }

        // Get in-block hit position and bias the result
        // a bit inside the clicked face, so we don't get ambiguous results.
        final var blockMiddle = block.getLocation().toVector().add(new Vector(0.5, 0.5, 0.5));
        final var hit = result
            .getHitPosition()
            .subtract(blockMiddle)
            .subtract(result.getHitBlockFace().getDirection().multiply(0.25));

        return new Oct(hit, result.getHitBlockFace());
    }

    public static class RaytraceDominantFaceResult {

        public BlockFace face = null;
        public double dominance = 0.0;
    }

    public static RaytraceDominantFaceResult raytraceDominantFace(final LivingEntity entity, final Block block) {
        // Ray trace clicked face
        final var result = entity.rayTraceBlocks(10.0);
        if (block == null || result == null || !block.equals(result.getHitBlock())) {
            return null;
        }

        final var blockMiddle = block.getLocation().toVector().add(new Vector(0.5, 0.5, 0.5));
        final var hitPosition = result.getHitPosition();
        final var diff = hitPosition.subtract(blockMiddle);

        final var ret = new RaytraceDominantFaceResult();
        for (final var face : BlockUtil.XZ_FACES) {
            // Calculate how dominant the current face contributes to the clicked point
            final double faceDominance;
            if (face.getModX() != 0) {
                faceDominance = diff.getX() * face.getModX();
            } else { // if (face.getModZ() != 0) {
                faceDominance = diff.getZ() * face.getModZ();
            }

            // Find maximum dominant face
            if (faceDominance > ret.dominance) {
                ret.face = face;
                ret.dominance = faceDominance;
            }
        }

        return ret;
    }

    public static String textureFromSkull(final Skull skull) {
        final var profile = skull.getProfile();
        if (profile == null) {
            return null;
        }

        for (final var property : profile.properties()) {
            if ("textures".equals(property.getName())) {
                return property.getValue();
            }
        }

        return null;
    }

    public static class BlockVectorRadiusComparator implements Comparator<BlockVector> {

        @Override
        public int compare(BlockVector a, BlockVector b) {
            return (
                (a.getBlockX() * a.getBlockX() + a.getBlockY() * a.getBlockY() + a.getBlockZ() * a.getBlockZ()) -
                (b.getBlockX() * b.getBlockX() + b.getBlockY() * b.getBlockY() + b.getBlockZ() * b.getBlockZ())
            );
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/util/BlockUtil.kt
package org.oddlama.vane.util

import net.minecraft.core.BlockPos
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.block.BlockFace
import org.bukkit.block.Skull
import org.bukkit.craftbukkit.CraftWorld
import org.bukkit.entity.LivingEntity
import org.bukkit.inventory.ItemStack
import org.bukkit.util.BlockVector
import org.bukkit.util.Vector

/**
 * Block and raytrace utility helpers.
 */
object BlockUtil {
    /** All six cardinal block faces. */
    @JvmField
    val BLOCK_FACES: List<BlockFace?> = listOf(
        BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH,
        BlockFace.WEST, BlockFace.UP, BlockFace.DOWN
    )

    /** Horizontal block faces on the XZ plane. */
    @JvmField
    val XZ_FACES: List<BlockFace> = listOf(
        BlockFace.NORTH, BlockFace.EAST, BlockFace.SOUTH, BlockFace.WEST
    )

    /** Maximum supported radius for precomputed nearest-block lookup. */
    const val NEAREST_RELATIVE_BLOCKS_FOR_RADIUS_MAX: Int = 6

    /** Precomputed nearest relative block vectors per radius. */
    @JvmField
    val NEAREST_RELATIVE_BLOCKS_FOR_RADIUS: List<List<BlockVector>> =
        (0..NEAREST_RELATIVE_BLOCKS_FOR_RADIUS_MAX).map { nearestBlocksForRadius(it) }

    /** Returns whether two blocks share the same coordinates. */
    fun equalsPos(b1: Block, b2: Block): Boolean =
        b1.x == b2.x && b1.y == b2.y && b1.z == b2.z

    /** Drops an item naturally at a block center. */
    @JvmStatic
    fun dropNaturally(block: Block, drop: ItemStack) =
        dropNaturally(block.location.add(0.5, 0.5, 0.5), drop)

    /** Drops an item naturally at a location with random position and velocity offsets. */
    @JvmStatic
    fun dropNaturally(loc: Location, drop: ItemStack) {
        loc.world.dropItem(
            loc.add(Vector.getRandom().subtract(Vector(.5, .5, .5)).multiply(0.5)), drop
        ).velocity = Vector.getRandom().add(Vector(-.5, +.5, -.5)).normalize().multiply(.15)
    }

    /** Returns nearby relative blocks in a radius, sorted by distance. */
    fun nearestBlocksForRadius(radius: Int): List<BlockVector> =
        buildList {
            for (x in -radius..radius) {
                for (z in -radius..radius) {
                    if (x * x + z * z <= radius * radius + 0.5) {
                        add(BlockVector(x, 0, z))
                    }
                }
            }
            sortWith(BlockVectorRadiusComparator())
        }

    /** Resolves a relative block from a root block and vector offset. */
    @JvmStatic
    fun relative(block: Block, relative: Vector): Block =
        block.getRelative(relative.blockX, relative.blockY, relative.blockZ)

    /** Finds the nearest tillable block around a root block. */
    @JvmStatic
    fun nextTillableBlock(rootBlock: Block, radius: Int, careless: Boolean): Block? {
        for (relativePos in NEAREST_RELATIVE_BLOCKS_FOR_RADIUS[radius]) {
            val block = relative(rootBlock, relativePos)
            if (!MaterialUtil.isTillable(block.type)) continue
            val above = block.getRelative(0, 1, 0)
            when {
                above.type == Material.AIR -> return block
                careless && MaterialUtil.isReplaceableGrass(above.type) -> {
                    above.type = Material.AIR
                    return block
                }
            }
        }
        return null
    }

    /** Returns direct, edge, and corner adjacent blocks around a root block. */
    @JvmStatic
    fun adjacentBlocks3D(root: Block): Array<Block?> = arrayOf(
        // Direct adjacent
        root.getRelative(1, 0, 0), root.getRelative(-1, 0, 0),
        root.getRelative(0, 0, 1), root.getRelative(0, 0, -1),
        root.getRelative(0, 1, 0), root.getRelative(0, -1, 0),
        // Edge adjacent
        root.getRelative(0, -1, -1), root.getRelative(0, -1, 1),
        root.getRelative(0, 1, -1), root.getRelative(0, 1, 1),
        root.getRelative(-1, 0, -1), root.getRelative(-1, 0, 1),
        root.getRelative(1, 0, -1), root.getRelative(1, 0, 1),
        root.getRelative(-1, -1, 0), root.getRelative(-1, 1, 0),
        root.getRelative(1, -1, 0), root.getRelative(1, 1, 0),
        // Corner adjacent
        root.getRelative(-1, -1, -1), root.getRelative(-1, -1, 1),
        root.getRelative(-1, 1, -1), root.getRelative(-1, 1, 1),
        root.getRelative(1, -1, -1), root.getRelative(1, -1, 1),
        root.getRelative(1, 1, -1), root.getRelative(1, 1, 1),
    )

    /** Finds the nearest air block above matching farmland around a root block. */
    @JvmStatic
    fun nextSeedableBlock(rootBlock: Block, farmlandType: Material?, radius: Int): Block? {
        for (relativePos in NEAREST_RELATIVE_BLOCKS_FOR_RADIUS[radius]) {
            val block = relative(rootBlock, relativePos)
            val below = block.getRelative(BlockFace.DOWN)
            if (below.type == farmlandType && block.type == Material.AIR) return block
        }
        return null
    }

    /** Triggers neighbor updates for a lever-like block state change. */
    @JvmStatic
    fun updateLever(block: Block, face: BlockFace) {
        val level = (block.world as CraftWorld).handle
        val connectedBlock = block.getRelative(face.oppositeFace)
        val blockPos1 = BlockPos(block.x, block.y, block.z)
        val blockPos2 = BlockPos(connectedBlock.x, connectedBlock.y, connectedBlock.z)
        val initiatorBlock = level.getBlockIfLoaded(blockPos1)!!
        level.updateNeighborsAt(blockPos1, initiatorBlock)
        level.updateNeighborsAt(blockPos2, initiatorBlock)
    }

    /** Raytraces a block hit and returns octant information. */
    @JvmStatic
    fun raytraceOct(entity: LivingEntity, block: Block?): Oct? {
        val result = entity.rayTraceBlocks(10.0) ?: return null
        if (block == null || block != result.hitBlock) return null

        val blockMiddle = block.location.toVector().add(Vector(0.5, 0.5, 0.5))
        val hit = result.hitPosition
            .subtract(blockMiddle)
            .subtract(result.hitBlockFace!!.direction.multiply(0.25))
        return Oct(hit, result.hitBlockFace)
    }

    /** Raytraces and returns the dominant horizontal face around the hit point. */
    @JvmStatic
    fun raytraceDominantFace(entity: LivingEntity, block: Block?): RaytraceDominantFaceResult? {
        val result = entity.rayTraceBlocks(10.0) ?: return null
        if (block == null || block != result.hitBlock) return null

        val blockMiddle = block.location.toVector().add(Vector(0.5, 0.5, 0.5))
        val diff = result.hitPosition.subtract(blockMiddle)

        val ret = RaytraceDominantFaceResult()
        for (face in XZ_FACES) {
            val faceDominance = if (face.modX != 0) diff.x * face.modX else diff.z * face.modZ
            if (faceDominance > ret.dominance) {
                ret.face = face
                ret.dominance = faceDominance
            }
        }
        return ret
    }

    /** Extracts a base64 texture payload from a skull profile. */
    @JvmStatic
    fun textureFromSkull(skull: Skull): String? =
        skull.profile?.properties()
            ?.firstOrNull { it.name == "textures" }
            ?.value

    /**
     * Encodes one of eight block-space corners.
     */
    class Corner private constructor(
        private val x: Boolean,
        private val y: Boolean,
        private val z: Boolean,
        private val id: Int
    ) {
        /** Creates a corner from explicit axis signs. */
        constructor(x: Boolean, y: Boolean, z: Boolean) : this(
            x, y, z,
            ((if (x) 1 else 0) shl 2) or ((if (y) 1 else 0) shl 1) or (if (z) 1 else 0)
        )

        /** Creates a corner from a hit vector relative to block center. */
        constructor(hit: Vector) : this(hit.x >= 0.0, hit.y >= 0.0, hit.z >= 0.0)

        /** Returns whether the corner is on the upper half. */
        fun up(): Boolean = y

        /** Returns whether the corner is on the eastern half. */
        fun east(): Boolean = x

        /** Returns whether the corner is on the southern half. */
        fun south(): Boolean = z

        /** Returns a copy with updated upper/lower half. */
        fun up(up: Boolean): Corner = Corner(x, up, z)

        /** Returns a copy with updated east/west half. */
        fun east(east: Boolean): Corner = Corner(east, y, z)

        /** Returns a copy with updated north/south half. */
        fun south(south: Boolean): Corner = Corner(x, y, south)

        /** Rotates this corner into north-facing reference orientation. */
        fun rotateToNorthReference(rotation: BlockFace): Corner = when (rotation) {
            BlockFace.NORTH -> Corner(x, y, z)
            BlockFace.EAST -> Corner(z, y, !x)
            BlockFace.SOUTH -> Corner(!x, y, !z)
            BlockFace.WEST -> Corner(!z, y, x)
            else -> throw IllegalArgumentException("rotation must be one of NORTH, EAST, SOUTH, WEST")
        }

        /** Returns the diagonal horizontal face represented by this corner. */
        fun xzFace(): BlockFace = when {
            x && z -> BlockFace.SOUTH_EAST
            x && !z -> BlockFace.NORTH_EAST
            !x && z -> BlockFace.SOUTH_WEST
            else -> BlockFace.NORTH_WEST
        }

        override fun hashCode(): Int = id
        override fun equals(other: Any?): Boolean = other is Corner && other.id == id
    }

    /**
     * Raytrace octant result.
     */
    class Oct(private val hitPos: Vector, private val face: BlockFace?) {
        /** Corner represented by this hit position. */
        val corner: Corner = Corner(hitPos)

        /** Returns hit position relative vector. */
        fun hitPos(): Vector = hitPos

        /** Returns hit corner. */
        fun corner(): Corner = corner

        /** Returns hit face. */
        fun face(): BlockFace? = face
    }

    /**
     * Result holder for dominant horizontal raytrace face checks.
     */
    class RaytraceDominantFaceResult {
        /** Dominant face candidate. */
        @JvmField
        var face: BlockFace? = null

        /** Dominance score for [face]. */
        @JvmField
        var dominance: Double = 0.0
    }

    /**
     * Comparator sorting block vectors by squared distance to origin.
     */
    class BlockVectorRadiusComparator : Comparator<BlockVector> {
        /** Compares vectors by squared Euclidean distance. */
        override fun compare(a: BlockVector, b: BlockVector): Int =
            (a.blockX * a.blockX + a.blockY * a.blockY + a.blockZ * a.blockZ) -
                    (b.blockX * b.blockX + b.blockY * b.blockY + b.blockZ * b.blockZ)
    }
}
